mirrorlist: add protocol option

adrianreber opened the pull request:
This adds the option specify the desired protocol of the mirrorlist/metalink as described in #100 

The code is currently running in staging and the result can be seen at:

http://mirrors.stg.fedoraproject.org/mirrorlist?repo=rawhide&arch=x86_64&country=jp&protocol=rsync
http://mirrors.stg.fedoraproject.org/mirrorlist?repo=rawhide&arch=x86_64&country=jp&protocol=ftp
http://mirrors.stg.fedoraproject.org/metalink?repo=rawhide&arch=x86_64&country=gb&protocol=https
http://mirrors.stg.fedoraproject.org/metalink?repo=rawhide&arch=x86_64&country=gb&protocol=rsync

pypingou commented on mirrorlist/mirrorlist_client.wsgi:
```diff
@@ -165,7 +165,7 @@ def application(environ, start_response):
         text += message + '\n'
         for (hostid, url) in results:
-            text += url + '\n'
+            text += url[0] + '\n'
```
This change surprises me a little

adrianreber replied:
This is described in the first commit. To trim the protocols to only contain the requested protocols I am re-using the function trim_to_preferred_protocols(). This function is currently called after the metalink is created and changes the result from [(hostid, [url, url, url]), ...] to [(hostid, url), ...], which was okay for the mirrorlist back-end as it only returned one URL per host. As the function trim_to_preferred_protocols() already reduces the URLs returned per host to a defined set (https, http, ftp) for metalink I wanted to re-use this function to trim the URL list to the requested protocol.

To be able to re-use trim_to_preferred_protocols() it needs to return a result which metalink and mirrorlist can handle. metalink wants multiple URLs per host and therefore I am now always returning [(hostid, [url, url, url]), ...] and [url, url, url] now always contains exactly one element for mirrorlist and therefore url[0].

pypingou replied:
ok thanks

pypingou commented:
:+1: for me

nirik commented:
Looks great! as a aside, we should likely ask mirrors to add more https links before we advertise this feaure. Or perhaps even scan all mirrors with http:// links and see how many support https and auto add those.

adrianreber commented:
Yes, scanning for https URLs should be possible and also easier than the ftp URLs removal.

adrianreber commented:
I scanned all http URLs and replaced 'http' with 'https' and added the result to the database. This increased the number of https URLs from 24 to 112. We have 920 http URLs in the database.

The metalink will start to redirect users to http and https URLs so people will be hitting more https mirrors in about 30 minutes.

I have not verified that the content provided via the same https URL is the same as the content from the http URL.
